src/service/template-impl.js

```diff
@@ -86,37 +86,70 @@ export class BaseTemplate {
   }
 
   /**
-   * Helps the template implementation to unwrap the root element. The root
-   * element can be unwrapped only when it contains a single element or a
-   * single element surrounded by empty text nodes.
+   * Iterate through the child nodes of the given root, applying the
+   * given callback to non-empty text nodes and elements.
    * @param {!Element} root
-   * @return {!Element}
-   * @protected @final
+   * @param {function((!Element|string))} callback
    */
-  unwrap(root) {
-    let singleElement = null;
+  visitChildren_(root, callback) {
     for (let n = root.firstChild; n != null; n = n.nextSibling) {
       if (n.nodeType == /* TEXT */ 3) {
-        if (n.textContent.trim()) {
-          // Non-empty text node - can't unwrap.
-          singleElement = null;
-          break;
+        const text = n.textContent.trim();
+        if (text) {
+          callback(text);
         }
       } else if (n.nodeType == /* COMMENT */ 8) {
         // Ignore comments.
       } else if (n.nodeType == /* ELEMENT */ 1) {
-        if (!singleElement) {
-          singleElement = dev().assertElement(n);
-        } else {
-          // This is not the first element - can't unwrap.
-          singleElement = null;
-          break;
-        }
+        callback(dev().assertElement(n));
+      }
+    }
+  }
+
+  /**
+   * Helps the template implementation to unwrap the root element. The root
+   * element can be unwrapped only when it contains a single element or a
+   * single element surrounded by empty text nodes.
+   * @param {!Element} root
+   * @return {!Element}
+   * @protected @final
+   */
+  tryUnwrap(root) {
+    let onlyChild;
+    this.visitChildren_(root, (c) => {
+      if (onlyChild === undefined && c.nodeType) {
+        onlyChild = c;
       } else {
-        singleElement = null;
+        onlyChild = null;
       }
+    });
+    return onlyChild || root;
+  }
+
+  /**
+   * Helps the template implementation to unwrap the root element. The root
+   * element contents are unwrapped as an array of elements. Any text node
+   * children are normalized inside a <div>. Any root with non-element children
+   * is returned as is to preserve the root tag rather than replaceed with <div>.
+   * @param {!Element} root
+   * @return {!Element|!Array<!Element>}
+   * @protected @final
+   */
+  unwrapChildren(root) {
+    const children = [];
+    if (!root.firstElementChild) {
```

Comment:
I double-checked the `<p>text</p>` case and `firstElementChild` is also `null`, so it would return `<p>text</p>`. I would consider it unexpected for `<p>text</p>` to return `<div>text</div>`, but because nothing relies on this behaving one way or the other, I don't feel too strongly either way. Removed.